prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -0,0 +1,171 @@
+    """
+    num_lines = len(lines)
+    filtered = lines.query("SUB_1 == 'NOT AVAILABLE' or SUB_2 == 'NOT AVAILABLE'")
+    num_filtered = len(filtered)
+    print(
+        f"dropping {num_filtered} lines with one or more substations listed as "
+        f"'NOT AVAILABLE' out of a starting total of {num_lines}"
+    )
+    return lines.query("SUB_1 != 'NOT AVAILABLE' and SUB_2 != 'NOT AVAILABLE'").copy()
+
+
+def filter_lines_with_no_matching_substations(lines, substations):
+    """Filter lines with one or more substation name not present in the ``substations``
+    data frame, and report the number dropped.
+
+    :param pandas.DataFrame lines: data frame of lines.
+    :param pandas.DataFrame substations: data frame of substations.
+    :return: (*pandas.DataFrame*) -- lines with matching substations.
+    """
+    num_lines = len(lines)
+    matching_names = substations["NAME"]  # noqa: F841
+    filtered = lines.query(
+        "SUB_1 not in @matching_names or SUB_2 not in @matching_names"
+    )
+    num_filtered = len(filtered)
+    print(
+        f"dropping {num_filtered} lines with one or more substations not found in "
+        f"substations table out of a starting total of {num_lines}"
+    )
+    return lines.query("SUB_1 in @matching_names and SUB_2 in @matching_names").copy()
+
+
+def filter_lines_with_nonmatching_substation_coords(lines, substations, threshold=100):
+    """Filter lines for which either the starting or ending substation, by name, has
+    coordinates judged as too far away (based on the ``threshold`` parameter) from the
+    coodinated listed for the line.
+
+    :param pandas.DataFrame lines: data frame of lines.
+    :param pandas.DataFrame substations: data frame of substations.
+    :param int/float threshold: maximum mismatch distance (miles).
+    :return: (*pandas.DataFrame*) -- lines with matching substations.
+    """
+
+    def find_distance_to_closest_start_substation(line, substations):
+        # Coordinates are initially (lon, lat); we reverse to (lat, lon) for haversine
+        start_coordinates = line.loc["COORDINATES"][0][::-1]
+        start_name = line.loc["SUB_1"]
+        matching_start_substations = substations.loc[substations.NAME == start_name]
+        minimum_start_distance = matching_start_substations.apply(
+            lambda x: haversine(start_coordinates, (x.LATITUDE, x.LONGITUDE)), axis=1
+        ).min()
+        return minimum_start_distance
+
+    def find_distance_to_closest_end_substation(line, substations):
+        # Coordinates are initially (lon, lat); we reverse to (lat, lon) for haversine
+        end_coordinates = line.loc["COORDINATES"][-1][::-1]
+        end_name = line.loc["SUB_2"]
+        matching_end_substations = substations.loc[substations.NAME == end_name]
+        minimum_end_distance = matching_end_substations.apply(
+            lambda x: haversine(end_coordinates, (x.LATITUDE, x.LONGITUDE)), axis=1
+        ).min()
+        return minimum_end_distance
```

Comment:
Maybe we can combine these two similar functions into one. Two options I could think of:
1. Add another optional parameter to specify `start` or `end`
2. Simply don't bother, compare both `start` and `end`, then return the maximum of the two. In this way, we can simplify the filtering logic below as well.

Reply:
I like this idea, I'll implement it.

Reply:
I implemented something similar: one function, different parameters.